ext/standard/basic_functions.stub.php

```diff
@@ -1393,6 +1393,8 @@ function boolval(mixed $value): bool {}
 
 function strval(mixed $value): string {}
 
+function is_initialized(object|string $scope, string $property) : bool {}
```

Comment:
maybe `property_is_initialized( object|string $object_or_class , string $property )` would make more sense as a name

is_initialized may be confused with other functions/expressions on variables such as is_null, isset, etc. `property_exists ( object|string $object_or_class , string $property ) : bool`

Also, my personal preference would be to have this work for untyped properties as well.
For untyped properties, it would be a bit shorter than https://www.php.net/manual/en/reflectionproperty.isinitialized.php to check if a property was initialized, and also work with dynamic properties.

- (e.g. allowing this to work for untyped properties would make distinguishing between $property->dynamicProperty being null and unset shorter)
- ReflectionProperty->isInitialized seems to permit dynamic and untyped properties

```
php > class TestClass { public $declared; public int $typed; }
php > $o = new TestClass();
php > unset($o->declared);
php > $o->dynamic = 'dynamic value';
php > var_export((new ReflectionProperty($o, 'dynamic'))->isInitialized($o));
true
php > var_export((new ReflectionProperty($o, 'declared'))->isInitialized($o));
false
```

Reply:
I hesitate to widen the scope so far for two reasons: First it means necessarily using handlers, which are going to be slower - not much better than just using the existing mechanism, in my opinion. Second, other than typed properties won't raise exceptions when you try to determine their state by normal means.

Reply:
Would it work to only use slow handlers if the property wasn't defined (i.e. instead of throwing, check the handlers first)?

I would have thought the fast path could be used for declared properties, and the slow path for dynamic properties, but I'm not 100% sure if there aren't any extra edge cases (e.g. public dynamic properties overriding a declared private property)

> Second, other than typed properties won't raise exceptions when you try to determine their state by normal means.

non-typed declared instance properties would emit notices when you were trying to differentiate between null and undefined by normal means - property_exists would return true for all declared properties even after unsetting them

Reply:
> I would have thought the fast path could be used for declared properties, and the slow path for dynamic properties, but I'm not 100% sure if there aren't any extra edge cases (e.g. public dynamic properties overriding a declared private property)

Indeed it would introduce edges, which I didn't consider when I said perf would be the reason I didn't want to use them ...

I didn't use them to keep the scope of the function narrow ...

The reflector works the way it does because it has no choice, precisely because it's using handlers. I think using handlers and so *having* to support dynamic and un-typed properties confuses the purpose of the function, the purpose of the function is very narrowly to check if a typed property is initialized in order to avoid exception *in normal code*.

Reply:
I'm aware of a possible problem with avoiding handlers ... an internal class may declare typed properties and handlers that do not use pre-allocated property slots.

I think it's acceptable, because any object doing that is not behaving correctly (and would seem to be wasting memory), and if/when found we probably should consider that a bug in the handlers.

If it turns out we must use handlers, I'd rather not have the function, because you already have reflection and there doesn't seem to be a win ...

Reply:
> I'm aware of a possible problem with avoiding handlers ... an internal class may declare typed properties and handlers that do not use pre-allocated property slots.
> 
> I think it's acceptable, because any object doing that is not behaving correctly (and would seem to be wasting memory), and if/when found we probably should consider that a bug in the handlers.

We've actually started doing this a lot recently, so that these properties end up in reflection.